library/qtest/src/Functions.qs

```diff
@@ -47,6 +47,26 @@ function RunAllTestCases<'T : Eq + Show>(test_cases : (String, () -> 'T, 'T)[])
     Mapped((name, case, result) -> TestCase(name, case, result), test_cases)
 }
 
+/// # Summary
+/// Given a function to test and an array of test cases of the form (input, expected_output), and a test mode, runs the test cases and returns the result of the test mode.
+///
+/// # Inputs
+/// - `test_suite_name` : A string representing the name of the test suite.
+/// - `func` : The function to test.
+/// - `test_cases` : An array of tuples of the form (input, expected_output).
+/// - `mode` : A function that takes an array of tuples of the form (test_name, test_case, expected_output) and returns a value of type 'U.
+///            Intended to be either `Qtest.Functions.CheckAllTestCases` or `Qtest.Functions.RunAllTestCases`.
+///
+/// # Example
+/// ```qsharp
+/// TestMatrix("Add One", x -> x + 1, [(2, 3), (3, 4)], CheckAllTestCases);
+/// ```
+
+function TestMatrix<'T, 'O : Show + Eq, 'U>(test_suite_name : String, func : 'T -> 'O, test_cases : ('T, 'O)[], mode : ((String, () -> 'O, 'O)[]) -> 'U) : 'U {
```

Comment:
I'm trying to understand the reasoning behind making `mode` such a flexible type. Is the user ever expected provide their own implementation here? If it can only ever be one of two well-known functions, maybe we can make it an "enum" ....... and by "enum" I mean hardcoded strings 😬  Okay maybe that's not much better either.

So I guess I'm asking, is `mode` typed this way because it's the best we can do with the current Q# (lacking enums, constants etc) or is there another reason?

Reply:
What if we split this into `RunTestMatrix` and `CheckTestMatrix`?

Reply:
It is typed this way because it is easy to pass in "CheckAll.." vs "RunAll.." and is also pluggable (users can provide their own implementation if desired). What if I maintained this flexibility but also exported some wrappers that just pass the `mode` in for the user, and called those wrappers `RunTestMatrix` and `CheckTestMatrix`? like this: [bf16c1e](https://github.com/microsoft/qsharp/pull/2037/commits/bf16c1efc6157436008bc3770f673b13c1664bb8)

Reply:
Yes I like that compromise. The simple use case should be simple, and I think the specific exports you added are good for that.